probe-rs: add initial support for black magic probe

xobs opened the pull request:
This adds initial support for using a Black Magic Probe with probe-rs.

This initial commit adds basic support for JTAG and SWD. The probe is operated using bit-bang commands, so it is very slow right now. JTAG in particular is operating bit-by-bit.

Future patches will increase performance.

```
$ cargo run -- info --probe 1d50:6018:97B6DC14
   Compiling probe-rs-tools v0.24.0 (/Users/seancross/Code/probe-rs/probe-rs-tools)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 2.55s
     Running `target/debug/probe-rs info --probe '1d50:6018:97B6DC14'`
Probing target via JTAG

ARM Chip with debug port Default:
Debug Port: DPv0, DP Designer: <unknown>
└── 0 MemoryAP (AmbaAhb3)
    └── ROM Table (Class 1), Designer: STMicroelectronics
        ├── Cortex-M4 SCS   (Generic IP component)
        │   └── CPUID
        │       ├── IMPLEMENTER: ARM Ltd
        │       ├── VARIANT: 0
        │       ├── PARTNO: Cortex-M4
        │       └── REVISION: 1
        ├── Cortex-M3 DWT   (Generic IP component)
        ├── Cortex-M3 FBP   (Generic IP component)
        ├── Cortex-M3 ITM   (Generic IP component)
        ├── Cortex-M4 TPIU  (Coresight Component)
        └── Cortex-M4 ETM   (Coresight Component)


RISC-V Chip:
  IDCODE: 0000000000
    Version:      0
    Part:         0
    Manufacturer: 0 (Unknown Manufacturer Code)
Error showing Xtensa chip information: An error originating from the DebugProbe occurred.

Caused by:
    Invalid instruction register access: 30

Probing target via SWD

ARM Chip with debug port Default:
Debug Port: DPv1, DP Designer: ARM Ltd
└── 0 MemoryAP (AmbaAhb3)
    └── ROM Table (Class 1), Designer: STMicroelectronics
        ├── Cortex-M4 SCS   (Generic IP component)
        │   └── CPUID
        │       ├── IMPLEMENTER: ARM Ltd
        │       ├── VARIANT: 0
        │       ├── PARTNO: Cortex-M4
        │       └── REVISION: 1
        ├── Cortex-M3 DWT   (Generic IP component)
        ├── Cortex-M3 FBP   (Generic IP component)
        ├── Cortex-M3 ITM   (Generic IP component)
        ├── Cortex-M4 TPIU  (Coresight Component)
        └── Cortex-M4 ETM   (Coresight Component)


Debugging RISC-V targets over SWD is not supported. For these targets, JTAG is the only supported protocol. RISC-V specific information cannot be printed.
Debugging Xtensa targets over SWD is not supported. For these targets, JTAG is the only supported protocol. Xtensa specific information cannot be printed.

$
```

chrysn commented:
Works like a charm for my nRF52 target, including 1h-long-running RTT sessions that have failed with other debuggers. As described it's a tad slow (noticeable only when programming, and then not abysmally slow-- maybe factor 10 compare to the STLink). Thanks for tackling this, it already helps my workflow!

xobs commented:
I've added the ability to batch JTAG commands, so up to 32-bits are now included in an single packet.

Additionally, I've added the ability to connect to a remote BMP such as an ESP32 running [Farpatch](https://github.com/farpatch/farpatch):

```
$ ./target/release/probe-rs info --protocol swd --probe 1d50:6018:10.0.237.165:2022
Probing target via SWD

ARM Chip with debug port Default:
Debug Port: DPv1, DP Designer: ARM Ltd
└── 0 MemoryAP (AmbaAhb3)
    └── ROM Table (Class 1), Designer: STMicroelectronics
        ├── Cortex-M4 SCS   (Generic IP component)
        │   └── CPUID
        │       ├── IMPLEMENTER: ARM Ltd
        │       ├── VARIANT: 0
        │       ├── PARTNO: Cortex-M4
        │       └── REVISION: 1
        ├── Cortex-M3 DWT   (Generic IP component)
        ├── Cortex-M3 FBP   (Generic IP component)
        ├── Cortex-M3 ITM   (Generic IP component)
        ├── Cortex-M4 TPIU  (Coresight Component)
        └── Cortex-M4 ETM   (Coresight Component)


Debugging RISC-V targets over SWD is not supported. For these targets, JTAG is the only supported protocol. RISC-V specific information cannot be printed.
Debugging Xtensa targets over SWD is not supported. For these targets, JTAG is the only supported protocol. Xtensa specific information cannot be printed.
```

xobs commented:
The latest version of this patch introduces full ADIv5 acceleration. It should be quite fast now, and should be considered feature complete.

The only thing missing is SWO which can be supported in a future patch if there is demand for it.

chrysn commented:
Somewhere down the dependency path there is a non-pinned dependency to something that is not semver compliant (or something that has a similar effect): Installing with `cargo install --locked --path probe-rs-tools` works, but without `--locked`, I got:

```
   Compiling capstone v0.12.0                                                                                                                                   
error[E0061]: this function takes 2 arguments but 3 arguments were supplied
   --> probe-rs-tools/src/bin/probe-rs/cmd/dap_server/debug_adapter/dap/repl_commands.rs:215:29                       
    |                                                                                                                                                           
215 |               let yaml_data = insta_yaml::serialize_value(
    |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^                 
216 |                   &target_core.core_data.stack_frames,
217 |                   insta_yaml::SerializationFormat::Yaml,
    |  ______________________________________________________-
218 | |                 insta_yaml::SnapshotLocation::File,
    | |                 ----------------------------------
    | |_________________|________________________________|
    |                   |                                help: remove the extra argument
    |                   unexpected argument of type `SnapshotLocation`          
    |                            
note: function defined here          
   --> /home/chrysn/.cargo/registry/src/index.crates.io-6f17d22bba15001f/insta-1.40.0/src/serialization.rs:91:8
    |                             
91  | pub fn serialize_value<S: Serialize>(s: &S, format: SerializationFormat) -> String {
    |        ^^^^^^^^^^^^^^^        
                                        
For more information about this error, try `rustc --explain E0061`.
```

The locked version builds with insta 1.39.0.

xobs commented:
I mentioned that in https://github.com/probe-rs/probe-rs/issues/2800#issuecomment-2336547115 but I thought it wasn't an issue since it's in `Cargo.lock`.

I wonder -- does `cargo install` ignore `Cargo.lock` by default? If so, then perhaps we should make a separate PR to pin that package to 1.39.0, or apply the patch I mention in that comment.

chrysn commented:
Well, it's at least unrelated to this PR, sorry for the noise.

At any rate, flashing is now really fast, thanks for the updates!

xobs commented:
No worries about the noise -- I really appreciate the testing, and am happy it's already useful for someone.

xobs commented:
I realize this is a relatively large patch, even though it's contained within its own new Probe. Would it be possible to get some feedback on it?

If this is not a desirable feature, then I can close the PR.

bugadani commented on probe-rs/src/probe/blackmagic/arm.rs:
```diff
@@ -0,0 +1,840 @@
+    }
+
+    fn write_8(&mut self, address: u64, data: &[u8]) -> Result<(), ArmError> {
+        self.write(0, address, data.as_bytes())
```
Can we avoid magic numbers here? `align = 0` doesn't have much meaning to me, I'm guessing this is some device-specific constant for 8-bit alignment. I'd prefer an `enum` that you can just `as u8` into the numeric value.

xobs replied:
Good idea, I'll do that.

bugadani commented on probe-rs/src/probe/blackmagic/mod.rs:
```diff
@@ -0,0 +1,1523 @@
+    }
+}
+
+impl<'a> core::fmt::Display for RemoteCommand<'a> {
```
Display is meant for human consumption and this looks more like the actual command encoding? I think it's fine to implement `ToString` instead.

xobs replied:
Coming from the embedded world, I still find it odd to be allocating strings all the time. But I agree that it's logically better to implement `ToString` rather than `Display`, so I'll go ahead and do that.

bugadani commented on probe-rs/src/probe/blackmagic/mod.rs:
```diff
@@ -0,0 +1,1523 @@
+}
+
+/// A Black Magic Probe.
+pub struct BlackMagicProbe<R: Read, W: Write> {
```
Can we use trait objects instead of the generics here?

xobs replied:
Good idea. Again, the concept of using heap-allocation is still somewhat novel. I've reworked it to use boxed traits, and the impl statements got much simpler.

chrysn commented:
Now that udev is disabled, `probe-rs list` does not find my BMP any more. All up to that commit works.

I've compared `--log-file` outputs before and after, but apparently that discovery is not logged.

Is there any output that'll help make the fallback discovery better? Or is there already some udev independent discovery in probe-rs that we can reuse to get the right USB device?

xobs commented:
Ultimately, what we want to have is a way to go from a USB VID/PID to a path. And the way to do that is system-dependent.

Let me see what I can do...

xobs commented:
I've reverted that change and added `libudev` back in. I'm trying to think if there's a different way to treat Linux...

chrysn commented:
I've skimmed through the code, and the FTDI probes (which I think depend on VID/PID detection as well) go through nusb in `probe-rs/src/probe/ftdi/mod.rs`. Might be that that would need to go into the serialport crate, though.

xobs commented:
Alright, with the latest patch I've removed the dependency on `libudev` and special case Linux. I'll see if I can't upstream these changes to serialport-rs.

chrysn commented on probe-rs/src/probe/blackmagic/mod.rs:
```diff
@@ -0,0 +1,1554 @@
+            let port_name = port_name.split('/').last()?;
+            let device_path = PathBuf::from(format!("/sys/class/tty/{}/device", port_name))
+                .canonicalize()
+                .ok()?;
+            let subsystem = device_path.join("subsystem").canonicalize().ok()?;
+            let subsystem = subsystem.file_name()?.to_string_lossy();
+
+            let usb_interface_path = if subsystem == "usb-serial" {
+                device_path.parent()?
+            } else if subsystem == "usb" {
+                &device_path
+            } else {
+                return None;
+            };
+            let usb_device_path = usb_interface_path.parent()?;
+
+            fn read_file_to_string(dir: &Path, file: &str) -> Option<String> {
+                let dir = dir.join(file);
+                let mut f = File::open(dir).ok()?;
+                let mut s = String::new();
+                f.read_to_string(&mut s).ok()?;
+                Some(s.trim().to_owned())
+            }
+
+            fn read_file_to_u16(dir: &Path, file: &str) -> Option<u16> {
+                u16::from_str_radix(&read_file_to_string(dir, file)?, 16).ok()
+            }
+
+            fn read_file_to_u8(dir: &Path, file: &str) -> Option<u8> {
+                u8::from_str_radix(&read_file_to_string(dir, file)?, 16).ok()
+            }
+
+            let vid = read_file_to_u16(&usb_device_path, &"idVendor")?;
+            let pid = read_file_to_u16(&usb_device_path, &"idProduct")?;
+            let interface_number = read_file_to_u8(&usb_interface_path, &"bInterfaceNumber");
+            let serial_number = read_file_to_string(&usb_device_path, &"serial");
+            let product_name = read_file_to_string(&usb_device_path, &"product")
+                .unwrap_or_else(|| "Black Magic Probe".to_string());
+            (vid, pid, serial_number, interface_number, product_name)
+        }
```
Is that the part that would be PR'd into serialport? If so, reviewers might appreciate a note right in the code; if that PR already exists, maybe even pointing to that PR so that it is easy to simplify the code again once serialport has improved.

(I don't think that having the workaround in here is detrimental to inclusion in probe-rs, especially if it is a clearly indicated workaround with a path forward, and the workaround is necessary due to probe-rs's policy not to use libudev).

xobs replied:
That's a good suggestion! I've added a comment to that effect and linked to the PR.

The new version isn't out yet, and there's no word on when it will be out. But once they release it, the workaround will continue to function and can be removed at a later date.

xobs commented:
I've fixed things that broke with the update to `zerocopy`, in addition to fixing clippy lints.

xobs commented:
Is there anything left to do? CI unfortunately doesn't complete because two jobs timed out after a day of not getting picked up.

xobs commented:
How would I do that? I'm following the steps in [CONTRIBUTING.md](https://github.com/probe-rs/probe-rs/blob/master/CONTRIBUTING.md) and it doesn't mention anything about how to add tags or who to tag.

bugadani commented:
Sorry, CI was broken because my setup fell apart. I guess we can now update to serialport 4.6?

okhsunrog commented:
@xobs I think you can ask in matrix: https://matrix.to/#/#probe-rs:matrix.org

Yatekii commented on probe-rs/src/probe/blackmagic/arm.rs:
```diff
@@ -0,0 +1,815 @@
+    }
+
+    fn current_debug_port(&self) -> DpAddress {
+        todo!()
```
This will lead to questionable UX. Can we find a way to implement this straight up? Or change the interface to return a Result worst case?

Yatekii commented on probe-rs/src/probe/blackmagic/arm.rs:
```diff
@@ -0,0 +1,815 @@
+}
+
+impl SwoAccess for BlackMagicProbeArmDebug {
+    fn enable_swo(&mut self, config: &crate::architecture::arm::SwoConfig) -> Result<(), ArmError> {
```
All of the trait methods should return an error (instead of panicking) (there should already be a `NotImplemented` variant) and not print anything :)

Yatekii replied:
Also, is the implementation of this trait really required? It should be optional I believe.

xobs replied:
Yes, it's required because `ArmProbeInterface` requires it to be implemented.

Yatekii replied:
Hmm that's suboptimal considering the fact that the methods seemingly are never called 😅

Yatekii replied:
They should still return `NotImplemented` instead of `unimplemented!()` to provide a nicer user experience :)

Yatekii commented on probe-rs/src/probe/blackmagic/arm.rs:
```diff
@@ -0,0 +1,815 @@
+        DpAddress::Default => Ok(0),
+        DpAddress::Multidrop(val) => val
+            .try_into()
+            .map_err(|_| ArmError::NotImplemented("multidrop value is out of range")),
```
NotImplemented is for functionality that's not implemented. If there is no fitting variant we can introduce one :)

xobs replied:
What should I return if it's supported but I haven't implemented it yet?

Yatekii replied:
Oh, it can support larger than u8 values? I guess I misunderstood then :)

Small nit: Can we make the error message tell the user what the allowed range of values is and what value was received? :)

xobs replied:
Sorry, I'd forgotten what this code was doing. It can only support `u8` values. If you had an extremely long JTAG chain with more than 256 TAPs, then this value would be exceeded. Would I use `OutOfBounds`?

Yatekii replied:
That sounds correct, let's go with it! Truth be told, I am not the most familiar with JTAG :)

Yatekii commented on probe-rs/src/probe/blackmagic/arm.rs:
```diff
@@ -0,0 +1,815 @@
+                addr,
+                value,
+            },
+            ProtocolVersion::V3 | ProtocolVersion::V4 => RemoteCommand::WriteApV3 {
```
Is there no `WriteApV2`? (just curious how the mappings came to happen :))

xobs replied:
The mappings are based on hapazard protocol extensions to the Black Magic Probe protocol. V2 did not change from V1, so it just uses the previous version.

This was largely compiled by referencing existing implementations at https://github.com/blackmagic-debug/blackmagic/tree/main/src/platforms/hosted/remote

Yatekii replied:
Oki, nice :) Thanks!

Yatekii commented on probe-rs/src/probe/blackmagic/arm.rs:
```diff
@@ -0,0 +1,815 @@
+    fn get_arm_communication_interface(
+        &mut self,
+    ) -> Result<
+        &mut crate::architecture::arm::ArmCommunicationInterface<
```
It would be amazing to shorten those types by having an import (I think they are used quite a bit) :)

Yatekii commented on probe-rs/src/probe/blackmagic/arm.rs:
```diff
@@ -0,0 +1,815 @@
+
+impl BlackMagicProbeMemoryInterface<'_> {
+    fn read_slice(&mut self, offset: u64, data: &mut [u8]) -> Result<(), ArmError> {
+        // Leave enough room for '&K' plus '#' plus an optional NULL.
```
For me as an uneducated person wrt BMP this comment is not fully informative. Could you provide some more info?

Yatekii commented on probe-rs/src/probe/blackmagic/mod.rs:
```diff
@@ -0,0 +1,1508 @@
+                4 => ProtocolVersion::V4,
+                _ => {
+                    return Err(DebugProbeError::ProbeCouldNotBeCreated(
+                        ProbeCreationError::Other("Unrecognized probe version response"),
```
I would love some more output to the user here: Something along the lines `Unrecognized probe version response: Allowed values are in range [0, 4] but the received value was {0}`.

xobs replied:
How do I do that? `ProbeCreationError` requires a `&'static str`, which means dynamic strings cannot be formatted.

xobs replied:
I extended `RemoteError` and I now return that.

Yatekii replied:
Thanks!

Yatekii commented on probe-rs/src/probe/blackmagic/mod.rs:
```diff
@@ -0,0 +1,1508 @@
+    }
+
+    fn select_protocol(&mut self, protocol: WireProtocol) -> Result<(), DebugProbeError> {
+        // if self.protocol == Some(protocol) {
+        //     return Ok(());
+        // }
```
```suggestion
```

Yatekii reviewed (commented):
Awesome stuff! I added some comments, but we can pretty much merge this :)

Maybe it would be nice to link (did I miss it?) the docs to the BMP protocol somehwere. Generally comments are a bit sparse, so for me as an outsider it will be rather hard to maintain this. But comments are scarce in general in probe-rs :(

dragonmux commented:
In regards the documentation aspect for the protocol, the BMD project is aware of the lack of documentation around this protocol and we are working to correct that. When such a document has been written and published, we will do a follow-up to Xobs' PR to introduce references and link out to the documentation that gets written to provide the relevant explanation and discuss the protocol's request-response structure and such. This is why the lack of reference to a specification or other form of documentation on the protocol.

Yatekii commented:
Oh, just to make it clear, I was not criticizing lack of BMP resources, but rather lack of links to such in the MR :) But I guess that's a direct outcome when there is a shortage of it already :)

xobs commented:
I think that addresses all of your cocnerns. Let me know if there's anything else!

Yatekii commented:
Thanks a lot! There is one open comment that I would like to see fixed (I added one more comment to hopefully precise what I would like to see :)) :)

xobs commented:
Alright, I made that change and rebased again.

Yatekii reviewed (approved):
Amazing, thanks a lot :) Great stuff :)

xobs commented:
Thanks for the merge!

Yatekii commented:
Good stuff will always be merged :) Sorry it took so long :(
